src/toil/job.py

```diff
@@ -851,6 +889,770 @@ def __del__(self):
+                # instantiated often {nlink, attemptNumber, total, cached, sigmaJob}
+                self.__dict__.update(stateDict)
+
+            @classmethod
+            def load(cls, fileName):
+                '''
+                Load the state of the cache from the cache state file (via the file handle provided)
+                :param str fileName: Path to the cache state file
+                '''
+                # Read the value from the cache state file then initialize and instance of
+                # _CacheState with it.
+                with open(fileName, 'r') as fH:
+                    cacheInfoDict = cPickle.load(fH)
+                return cls(cacheInfoDict)
+
+            def write(self, fileName):
+                '''
+                Write the current state of the cache into a temporary file then atomically rename it
+                to the main cache state file.
+                :param str fileName: Path to the cache state file
+                :return:
+                '''
+                with open(fileName + '.tmp', 'w') as fH:
+                    # Based on answer by user "Mark" at:
+                    # http://stackoverflow.com/questions/2709800/how-to-pickle-yourself
+                    # We can't pickle nested classes. So we have to pickle the variables of the class
+                    cPickle.dump(self.__dict__, fH, cPickle.HIGHEST_PROTOCOL)
+                os.rename(fileName + '.tmp', fileName)
+
+
+            def isBalanced(self):
+                '''
+                Checks for the inequality of the caching equation, i.e.
+                                cachedSpace + sigmaJobDisk <= totalFreeSpace
+                Essentially, the sum of all cached file + disk requirements of all running jobs
+                should always be less than the available space on the system
+                :return: Boolean for equation is balanced (T) or not (F)
+                '''
+                return self.cached + self.sigmaJob <= self.total
+
+        def _accountForNlinkEquals2(self, localFilePath):
+            '''
+            This is a utility function that accounts for the fact that if nlinkThreshold == 2, the
+            size fo the file is accounted for by the file store copy of the file and thus the file
+            size shouldn't be added to the cached file sizes.
+            :param str localFilePath: Path to the local file that was linked to the file store copy.
+            :return: None
+            '''
+            fileStats = os.stat(localFilePath)
+            assert fileStats.st_nlink >= self.nlinkThreshold
+            with self.cacheLock() as lockFileHandle:
+                cacheInfo = self._CacheState.load(self.cacheStateFile)
+                cacheInfo.sigmaJob -= fileStats.st_size
+                self._updateJobReqs(cacheInfo, fileStats.st_size)
+                cacheInfo.write(self.cacheStateFile)
+
```

Comment:
Okay, so I'm starting to understand this. You have Job.Filestore, which implements my first version of caching and you have Job.CacheFileStore (or whatever it's called), that inherits from Job.FileStore. I think we should just replace Job.FileStore with Job.CacheFileStore. It will save about lines of 300 code, I think.